File: tasks/money-rollup/python/solution/money.py
"""Exact decimal helpers for money amounts and rates."""

import os
import re
from fractions import Fraction

AMOUNT = re.compile(r"^-?[0-9]+(\.[0-9]+)?$")
RATE = re.compile(r"^[0-9]+(\.[0-9]+)?$")


def parse_decimal(text, pattern, max_places):
    if not isinstance(text, str) or not pattern.match(text):
        raise ValueError("malformed decimal")
    body = text[1:] if text.startswith("-") else text
    places = len(body.split(".")[1]) if "." in body else 0
    if places > max_places and os.environ.get("LAB_SABOTAGE") != "ignore-decimal-limit":
        raise ValueError("too many decimal places")
    value = Fraction(int(body.replace(".", "")), 10**places)
    return -value if text.startswith("-") else value


def parse_amount(text, minor_units):
    return parse_decimal(text, AMOUNT, minor_units)


def parse_rate(text):
    value = parse_decimal(text, RATE, 8)
    if value <= 0:
        raise ValueError("rate must be positive")
    return value


def round_half_even(value, places):
    scaled = value * 10**places
    negative = scaled < 0
    magnitude = -scaled if negative else scaled
    whole = magnitude.numerator // magnitude.denominator
    remainder = magnitude - whole
    half_up = os.environ.get("LAB_SABOTAGE") == "half-up-rounding"
    if remainder > Fraction(1, 2):
        whole += 1
    elif remainder == Fraction(1, 2) and (half_up or whole % 2 == 1):
        whole += 1
    return -whole if negative else whole


def format_minor(minor, places):
    negative = minor < 0
    digits = str(-minor if negative else minor).rjust(places + 1, "0")
    if places:
        digits = digits[: len(digits) - places] + "." + digits[len(digits) - places :]
    return "-" + digits if negative else digits

File: tasks/money-rollup/javascript/solution/money.js
"use strict";

const AMOUNT = /^-?[0-9]+(\.[0-9]+)?$/;
const RATE = /^[0-9]+(\.[0-9]+)?$/;

function rational(numerator, denominator) {
  return { numerator, denominator };
}

function multiply(left, right) {
  return rational(
    left.numerator * right.numerator,
    left.denominator * right.denominator
  );
}

function parseDecimal(text, pattern, maxPlaces) {
  if (typeof text !== "string" || !pattern.test(text)) {
    throw new Error("malformed decimal");
  }
  const negative = text.startsWith("-");
  const body = negative ? text.slice(1) : text;
  const parts = body.split(".");
  const fraction = parts.length > 1 ? parts[1] : "";
  const ignoreLimit = process.env.LAB_SABOTAGE === "ignore-decimal-limit";
  if (fraction.length > maxPlaces && !ignoreLimit) {
    throw new Error("too many decimal places");
  }
  const digits = BigInt(parts[0] + fraction);
  return rational(negative ? -digits : digits, 10n ** BigInt(fraction.length));
}

function parseAmount(text, minorUnits) {
  return parseDecimal(text, AMOUNT, minorUnits);
}

function parseRate(text) {
  const value = parseDecimal(text, RATE, 8);
  if (value.numerator <= 0n) {
    throw new Error("rate must be positive");
  }
  return value;
}

function roundHalfEven(value, places) {
  const scaledNumerator = value.numerator * 10n ** BigInt(places);
  const negative = scaledNumerator < 0n;
  const numerator = negative ? -scaledNumerator : scaledNumerator;
  const denominator = value.denominator;
  let whole = numerator / denominator;
  const twiceRemainder = (numerator % denominator) * 2n;
  const halfUp = process.env.LAB_SABOTAGE === "half-up-rounding";
  const tie = twiceRemainder === denominator;
  if (twiceRemainder > denominator || (tie && (halfUp || whole % 2n === 1n))) {
    whole += 1n;
  }
  return negative ? -whole : whole;
}

function formatMinor(minor, places) {
  const negative = minor < 0n;
  let digits = (negative ? -minor : minor).toString().padStart(places + 1, "0");
  if (places) {
    const split = digits.length - places;
    digits = digits.slice(0, split) + "." + digits.slice(split);
  }
  return negative ? "-" + digits : digits;
}

module.exports = {
  formatMinor,
  multiply,
  parseAmount,
  parseRate,
  rational,
  roundHalfEven,
};
